python/pycylon/pycylon/data/table.pyx

```diff
@@ -1197,6 +1197,56 @@ cdef class Table:
             final_table = Table.from_list(self.context, col_names, filtered_all_data)
             return final_table
 
+    cdef _get_head_response(self, offset, length):
+        cdef shared_ptr[CTable] output
+        cdef CStatus status
+        status = Head(self.table_shd_ptr, length, &output)
+
+        if status.is_ok():
+            return pycylon_wrap_table(output)
+        else:
+            raise ValueError(f"Head operation failed : {status.get_msg().decode()}")
+
+    cdef _get_tail_response(self, offset, length):
+        cdef shared_ptr[CTable] output
+        cdef CStatus status
+        status = Tail(self.table_shd_ptr, length, &output)
+
+        if status.is_ok():
+            return pycylon_wrap_table(output)
+        else:
+            raise ValueError(f"Tail operation failed : {status.get_msg().decode()}")
+
+    cdef _get_distributed_slice_response(self, offset, length):
+        cdef shared_ptr[CTable] output
+        cdef CStatus status
+        status = DistributedSlice(self.table_shd_ptr, offset, length, &output)
+
+        if status.is_ok():
+            return pycylon_wrap_table(output)
+        else:
+            raise ValueError(f"DistributedSlice operation failed : {status.get_msg().decode()}")
+
+    cdef _get_distributed_head_response(self, offset, length):
+        cdef shared_ptr[CTable] output
+        cdef CStatus status
+        status = DistributedHead(self.table_shd_ptr, length, &output)
+
+        if status.is_ok():
+            return pycylon_wrap_table(output)
+        else:
+            raise ValueError(f"DistributedHead operation failed : {status.get_msg().decode()}")
```

Comment:
this logic is redundant over the three functions, it is better to write generic function which accepts a function pointer along with the arguments and just call the generic function as needed.

Reply:
Correct. Initially, it was in a single function. But later it was decided by discussing with Niranda multiple functions by design whether we can create a callback from cython to execute a function pointer in cpp layer. Those are the individual function. But it is a limitation at cython layer that we can't call function pointer.

Reply:
What do you mean by a limitation?

Reply:
Code
-------

```python
cdef add(arg1, arg2):
    return arg1 + arg2

cdef subtract(arg1, arg2):
    return arg1 - arg2

cdef multiply(arg1, arg2):
    return arg1 * arg2

cdef divide(arg1, arg2):
    return arg1 / arg2

cdef call_op(func_ptr, arg1, arg2):
    return func_ptr(arg1, arg2)


def run_ops(arg1, arg2, op="add"):
    """
    >>> from helloworld import run_ops
    >>> run_ops(10,20, "add")
    30

    """
    if op == "add":
        return call_op(add, arg1, arg2)
    elif op == "subtract":
        return call_op(subtract, arg1, arg2)
    elif op == "multiply":
        return call_op(multiply, arg1, arg2)
    elif op == "divide":
        return call_op(divide, arg1, arg2)
    else:
        raise ValueError(f"Unsupported math op {op}")

```





Example
--------

```python
>>> from helloworld import run_ops
>>> run_ops(10, 20, "add")
30
>>> run_ops(10, 20, "subtract")
-10
>>> run_ops(10, 20, "multiply")
200
>>> run_ops(10, 20, "divide")
0.5
>>> run_ops(10, 20, "hash")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "helloworld.pyx", line 34, in helloworld.run_ops
    raise ValueError(f"Unsupported math op {op}")
ValueError: Unsupported math op hash
```

Reply:
That's what I did in the first commit: b5b547075d356bd252dc1eb09fe56ebdfe50d34b
Please check it.

Later I changed it after communicating with Niranda for making separate handler functions.

Reply:
> What do you mean by a limitation?

You can't create a function pointer inside cython binding.

Reply:
Submitted new changes @vibhatha